### DISCUSSION TITLE
prevent double call of:  xTaskCreatePinnedToCore

### DISCUSSION THREAD

**Discussion by rubenfrolic** [espressif/arduino-esp32#7758] [Q&A]
In an application playing some LED animations on the second core in parallel with the rest of the application. I'm doing this by creating a task using `xTaskCreatePinnedToCore`, this task plays the animation, and when finished the tasks deletes it self: `vTaskDelete(NULL)` and then the esp will be put into deepsleep. 

It might happen that the main application will try to start an animation, while the other one is still playing. This results in `xTaskCreatePinnedToCore` being called, while the task is still running. I'd like to prevent this by checking the state of the task before I create it again. 

conceptually I wanted to do something like this:

```
 eTaskState ts = eTaskGetState(LEDS);

if(ts != eRunning){

  xTaskCreatePinnedToCore(
    updateLed, /* Task function. */
    "LEDS",    /* name of task. */
    10000,     /* Stack size of task */
    NULL,      /* parameter of the task */
    1,         /* priority of the task */
    &LEDS, /* Task handle to keep track of created task */
    0);    /* pin task to core 0 */

}
```
This unfortunately is not working and reboots the module if there's no task being created yet:

`assert failed: eTaskGetState tasks.c:1609 (pxTCB)`

I basically have two questions:

Is there any harm in calling `xTaskCreatePinnedToCore` while the task is already running?
Is there another way to check if a task is already running without getting into the problem above?

**Comment by mrengineer7777** [comment]
>Is there any harm in calling xTaskCreatePinnedToCore while the task is already running?

Yes, it will create another thread running the task.

>Is there another way to check if a task is already running without getting into the problem above?

Haven't done this yet, but as I understand it there are multiple options:
1) Check if LEDS is null.  Set `LEDS=0` before exiting task.  There may also be a way to capture the exception thrown by eTaskGetState or use a different way to check the task.  Note tasks are handled by the FreeRTOS operating system, but not all task diagnostics are compiled for Arduino core.
2) Use a semaphore or mutex to delay/abort task if it is already running.
3) Start task once and leave it running.  Use an event que and have the task wait (pause) until an event is put into the que.  This is the way several services are constructed in the Arduino libraries, and the approach I will likely take when I convert my LED controller library (serial) to real-time.

**Comment by BillyGriffiths** [comment]
You could also just check the state of the task by defining the task handle globally...

TaskHandle_t taskHandleMyTask = NULL;

void taskMyTask(void *params) {
    // Do something

    // Once done, the task deletes itself
    taskHandleMyTask = NULL;
    vTaskDelete(taskHandleMyTask);
}


void runMyTask() {

    // If the task is already running, do not start a new one
    if (taskHandleMyTask != NULL) {
        return;
    }

    // Create a "run-once" task
    xTaskCreate(taskMyTask, "taskMyTask", 2048, (void *)1, 1, &taskHandleMyTask);
}
